Sync/Synchronizers/ClientsSynchronizer.swift

```diff
@@ -398,9 +402,44 @@ open class ClientsSynchronizer: TimestampedSingleCollectionSynchronizer, Synchro
         // clients since the beginning of time instead of looking at `self.lastFetched`.
         return clientsClient.getSince(0)
             >>== { response in
-                return self.wipeIfNecessary(localClients)
-                    >>> { self.applyStorageResponse(response, toLocalClients: localClients, withServer: clientsClient, notifier: notifier) }
+                return self.maybeDeleteClients(response: response, withServer: storageClient)
+                    >>> { self.wipeIfNecessary(localClients)
+                                >>> { self.applyStorageResponse(response, toLocalClients: localClients, withServer: clientsClient, notifier: notifier) }
+                    }
+                
             }
             >>> { deferMaybe(self.completedWithStats) }
     }
+    
+    func maybeDeleteClients(response: StorageResponse<[Record<ClientPayload>]>, withServer storageClient: Sync15StorageClient) -> Success {
+        let hasOldClientGuidValue = response.value.contains { $0.id == self.scratchpad.clientGUID }
+        self.hasLocalClientIdChanged = false
+
+        if hasOldClientGuidValue {
+            return storageClient.deleteObject(collection: "clients", guid: self.scratchpad.clientGUID).bind { result in
+                if result.isSuccess {
+                    self.hasLocalClientIdChanged = true
+                    return succeed()
+                } else if result.failureValue is NotFound<HTTPURLResponse> {
+                    self.hasLocalClientIdChanged = true
```

Comment:
@mhammond At this point in the code, I've retrieved all the clients from the server and found that one of them has an ID with a value equal to the old client GUID, so I've made the call to delete the client from the server and gotten a 404 in return. At present I'm setting `self.hasLocalClientIdChanged` to true which will force a new client record that has an ID with the value of the local FxA Device ID to be uploaded to the server.

This seems reasonable to me because it's possible that a wipe was executed after all the client records were retrieved from the server. But maybe it's heavy handed as I'm not sure how likely that would be. Anyway, I'm raising it here just in case my approach should be reconsidered.

Here I am forcing the re-upload to happen, but I'm not sure if that's necessary. It might be enough to not treat this as an error. However if the client has been wiped since we retrieved all of our clients, it's perhaps possible that we still need to upload a local client record.

Reply:
This does seem wrong - it *should* be impossible to first find the guid in the response, but then get a 404 deleting it - I can't even think of a race condition that would make it necessary. I suspect the delete isn't working for some reason? Now about:sync is back alive, you should be able to verify the old record has died on the first sync after this patch is in place.

Reply:
I've actually done this in my testing. After the delete call I have verified in about:sync that the client record has been deleted leaving an orphaned tabs record (which isn't displayed in the synced tabs of any of my devices). So I think I'm going to remove the if clause for 404 so that it's treated as an error case in this scenario.